File: app/src/main/java/com/example/david/lists/data/remote/SnapshotListener.java
package com.example.david.lists.data.remote;

import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.example.david.lists.data.datamodel.Item;
import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.data.repository.IRepositoryContract;
import com.example.david.lists.util.UtilExceptions;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentChange;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.MetadataChanges;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;

import static com.example.david.lists.data.remote.RemoteRepositoryConstants.FIELD_ITEM_LIST_ID;
import static com.example.david.lists.data.remote.RemoteRepositoryConstants.FIELD_POSITION;

public final class SnapshotListener implements IRemoteRepositoryContract.SnapshotListener {

    private final Flowable<List<UserList>> userListFlowable;
    private final CollectionReference userListCollection;

    private final CollectionReference itemCollection;

    private ListenerRegistration userListsSnapshotListener;
    private ListenerRegistration itemsSnapshotListener;

    private final Flowable<List<UserList>> deletedUserListsFlowable;
    private FlowableEmitter<List<UserList>> deletedUserListsEmitter;

    private boolean recentLocalChanges;

    public SnapshotListener(CollectionReference userListCollection,
                            CollectionReference itemCollection,
                            IRepositoryContract.UserRepository userRepository,
                            FirebaseFirestore firestore) {
        this.deletedUserListsFlowable = initDeletedUserListsFlowable();
        this.userListFlowable = initUserListFlowable();
        this.userListCollection = userListCollection;
        this.itemCollection = itemCollection;
        recentLocalChanges = false;

        initFirebaseAuth(userRepository, firestore);
    }

    private Flowable<List<UserList>> initDeletedUserListsFlowable() {
        return Flowable.create(
                emitter -> this.deletedUserListsEmitter = emitter,
                BackpressureStrategy.BUFFER
        );
    }

    private Flowable<List<UserList>> initUserListFlowable() {
        return Flowable.create(
                this::userListQuerySnapshot,
                BackpressureStrategy.BUFFER
        );
    }


    @Override
    public Flowable<List<UserList>> getUserListFlowable() {
        return userListFlowable;
    }

    @Override
    public Flowable<List<Item>> getItemFlowable(String userListId) {
        return initItemFlowable(userListId);
    }

    @Override
    public Flowable<List<UserList>> getDeletedUserLists() {
        return deletedUserListsFlowable;
    }


    private void initFirebaseAuth(IRepositoryContract.UserRepository userRepository, FirebaseFirestore firestore) {
        userRepository.userSignedOutObservable().observeForever(signedOut -> {
            if (!signedOut) {
                return;
            }

            if (userListsSnapshotListener != null) {
                userListsSnapshotListener.remove();
            }
            if (itemsSnapshotListener != null) {
                itemsSnapshotListener.remove();
            }

            // Clear persistence storage
            // https://firebase.google.com/support/release-notes/android#version_2010
            firestore.clearPersistence();
        });
    }


    private void userListQuerySnapshot(FlowableEmitter<List<UserList>> emitter) {
        // this saves a reference to the listener
        this.userListsSnapshotListener = userListCollection
                .orderBy(FIELD_POSITION, Query.Direction.ASCENDING)
                .addSnapshotListener(MetadataChanges.INCLUDE, getUserListSnapshotListener(emitter));

        emitter.setCancellable(() -> {
            if (userListsSnapshotListener != null) {
                userListsSnapshotListener.remove();
            }
        });
    }

    private EventListener<QuerySnapshot> getUserListSnapshotListener(FlowableEmitter<List<UserList>> emitter) {
        return (queryDocumentSnapshots, e) -> {
            if (queryDocumentSnapshots == null) {
                return;
            }

            if (checkForErrorFromQuery(queryDocumentSnapshots, e)) {
                UtilExceptions.throwException(e);
                emitter.onError(e);
            } else if (shouldReturn(emitter, queryDocumentSnapshots)) {
                return;
            }

            if (deletedUserListsHasSubscribers()) {
                checkIfUserListDeleted(queryDocumentSnapshots);
            }

            emitter.onNext(queryDocumentSnapshots.toObjects(UserList.class));
        };
    }

    private boolean deletedUserListsHasSubscribers() {
        return deletedUserListsEmitter != null && !deletedUserListsEmitter.isCancelled();
    }

    private void checkIfUserListDeleted(QuerySnapshot queryDocumentSnapshots) {
        List<UserList> deletedUserLists = new ArrayList<>();
        for (DocumentChange change : queryDocumentSnapshots.getDocumentChanges()) {
            if (change.getType() == DocumentChange.Type.REMOVED) {
                deletedUserLists.add(change.getDocument().toObject(UserList.class));
            }
        }
        if (deletedUserLists.isEmpty()) {
            return;
        }
        this.deletedUserListsEmitter.onNext(deletedUserLists);
    }


    private Flowable<List<Item>> initItemFlowable(String userListId) {
        return Flowable.create(
                emitter -> itemQuerySnapshot(emitter, userListId),
                BackpressureStrategy.BUFFER
        );
    }

    private void itemQuerySnapshot(FlowableEmitter<List<Item>> emitter, String userListId) {
        this.itemsSnapshotListener = itemCollection
                .whereEqualTo(FIELD_ITEM_LIST_ID, userListId)
                .orderBy(FIELD_POSITION, Query.Direction.ASCENDING)
                .addSnapshotListener(MetadataChanges.INCLUDE, getItemSnapshotListener(emitter));

        emitter.setCancellable(() -> {
            if (itemsSnapshotListener != null) {
                itemsSnapshotListener.remove();
            }
        });
    }

    private EventListener<QuerySnapshot> getItemSnapshotListener(FlowableEmitter<List<Item>> emitter) {
        return (queryDocumentSnapshots, e) -> {
            if (queryDocumentSnapshots == null) {
                return;
            }

            if (checkForErrorFromQuery(queryDocumentSnapshots, e)) {
                UtilExceptions.throwException(e);
                emitter.onError(e);
            } else if (shouldReturn(emitter, queryDocumentSnapshots)) {
                return;
            }

            emitter.onNext(queryDocumentSnapshots.toObjects(Item.class));
        };
    }


    private boolean checkForErrorFromQuery(QuerySnapshot queryDocumentSnapshots, FirebaseFirestoreException e) {
        if (e != null) {
            return true;
        } else if (queryDocumentSnapshots == null) {
            Crashlytics.log(Log.ERROR, RemoteRepository.class.getSimpleName(), "QueryDocumentSnapshot is null");
            return true;
        }
        return false;
    }

    private boolean shouldReturn(FlowableEmitter emitter, QuerySnapshot queryDocumentSnapshots) {
        if (emitter.isCancelled()) {
            return true;
        }

        if (isRecentLocalChanges()) {
            recentLocalChanges = false;
            return true;
        } else if (fromLocalCache(queryDocumentSnapshots)) {
            recentLocalChanges = true;
            return false;
        }
        return false;
    }

    /**
     * Because there were recent local changes, I can assume that this payload is from the server
     * - which is identical of the query that just came from the local cache - thus it can be skipped.
     */
    private boolean isRecentLocalChanges() {
        return recentLocalChanges;
    }

    /**
     * This payload is from the local cache, post a local change.
     */
    private boolean fromLocalCache(QuerySnapshot queryDocumentSnapshots) {
        return queryDocumentSnapshots.getMetadata().hasPendingWrites();
    }
}

File: app/src/main/java/com/example/david/lists/data/remote/SnapshotListener.kt
package com.example.david.lists.data.remote

import android.util.Log
import com.crashlytics.android.Crashlytics
import com.example.david.lists.data.datamodel.Item
import com.example.david.lists.data.datamodel.UserList
import com.example.david.lists.data.remote.RemoteRepositoryConstants.FIELD_ITEM_USER_LIST_ID
import com.example.david.lists.data.remote.RemoteRepositoryConstants.FIELD_POSITION
import com.example.david.lists.data.repository.IRepositoryContract
import com.example.david.lists.util.UtilExceptions
import com.google.firebase.firestore.*
import com.google.firebase.firestore.DocumentChange.Type.REMOVED
import com.google.firebase.firestore.EventListener
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter
import java.util.*

class SnapshotListener(private val userListCollection: CollectionReference,
                       private val itemCollection: CollectionReference,
                       userRepo: IRepositoryContract.UserRepository,
                       firestore: FirebaseFirestore) : IRemoteRepositoryContract.SnapshotListener {

    override val userListFlowable: Flowable<List<UserList>>

    private var userListsSnapshotListener: ListenerRegistration? = null
    private var itemsSnapshotListener: ListenerRegistration? = null

    override val deletedUserListsFlowable: Flowable<List<UserList>>
    private var deletedUserListsEmitter: FlowableEmitter<List<UserList>>? = null

    /**
     * Because there were recent local changes, I can assume that this payload is from the server
     * - which is identical of the query that just came from the local cache - thus it can be skipped.
     */
    private var recentLocalChanges = false

    init {
        deletedUserListsFlowable = initDeletedUserListsFlowable()
        userListFlowable = initUserListFlowable()

        initFirebaseAuth(userRepo, firestore)
    }

    private fun initDeletedUserListsFlowable() =
            Flowable.create<List<UserList>>(
                    { deletedUserListsEmitter = it },
                    BackpressureStrategy.BUFFER
            )

    private fun initUserListFlowable() =
            Flowable.create<List<UserList>>(
                    { getUserListQuerySnapshot(it) },
                    BackpressureStrategy.BUFFER
            )

    private fun initFirebaseAuth(userRepo: IRepositoryContract.UserRepository, firestore: FirebaseFirestore) {
        userRepo.userSignedOutObservable().observeForever { signedOut ->
            if (signedOut) {
                userListsSnapshotListener?.remove()
                itemsSnapshotListener?.remove()
                firestore.clearPersistence()
            }
        }
    }


    private fun getUserListQuerySnapshot(emitter: FlowableEmitter<List<UserList>>) {
        this.userListsSnapshotListener = userListCollection
                .orderBy(FIELD_POSITION, Query.Direction.ASCENDING)
                .addSnapshotListener(MetadataChanges.INCLUDE, getUserListEventListener(emitter))

        emitter.setCancellable {
            userListsSnapshotListener?.remove()
        }
    }

    private fun getUserListEventListener(emitter: FlowableEmitter<List<UserList>>) =
            EventListener<QuerySnapshot> { querySnapshot, e ->
                if (validQuery(querySnapshot, e, emitter)) {
                    emitter.onNext(querySnapshot!!.toObjects(UserList::class.java))
                }

                if (deletedUserListsHasSubscribers()) {
                    checkIfUserListDeleted(querySnapshot!!)
                }
            }

    private fun deletedUserListsHasSubscribers() =
            deletedUserListsEmitter?.isCancelled?.not() ?: false

    private fun checkIfUserListDeleted(querySnapshot: QuerySnapshot) {
        val deletedUserLists = ArrayList<UserList>()
        for (change in querySnapshot.documentChanges) {
            if (change.type == REMOVED) {
                deletedUserLists.add(change.document.toObject(UserList::class.java))
            }
        }

        if (deletedUserLists.isEmpty()) {
            return
        }

        deletedUserListsEmitter!!.onNext(deletedUserLists)
    }


    override fun getItemFlowable(userListId: String) =
            Flowable.create<List<Item>>(
                    { getItemQuerySnapshot(it, userListId) },
                    BackpressureStrategy.BUFFER
            )

    private fun getItemQuerySnapshot(emitter: FlowableEmitter<List<Item>>, userListId: String) {
        itemsSnapshotListener = itemCollection
                .whereEqualTo(FIELD_ITEM_USER_LIST_ID, userListId)
                .orderBy(FIELD_POSITION, Query.Direction.ASCENDING)
                .addSnapshotListener(MetadataChanges.INCLUDE, getItemEventListener(emitter))

        emitter.setCancellable {
            itemsSnapshotListener?.remove()
        }
    }

    private fun getItemEventListener(emitter: FlowableEmitter<List<Item>>) =
            EventListener<QuerySnapshot> { querySnapshot, e ->
                if (validQuery(querySnapshot, e, emitter)) {
                    emitter.onNext(querySnapshot!!.toObjects(Item::class.java))
                }
            }


    private fun validQuery(querySnapshot: QuerySnapshot?,
                           e: FirebaseFirestoreException?,
                           emitter: FlowableEmitter<*>): Boolean {
        if (querySnapshot === null) {
            return false
        }

        if (queryError(querySnapshot, e)) {
            UtilExceptions.throwException(e!!)
            emitter.onError(e)
            return false
        } else if (shouldReturn(emitter, querySnapshot)) {
            return false
        }
        return true
    }

    private fun queryError(querySnapshot: QuerySnapshot?, e: FirebaseFirestoreException?): Boolean {
        if (e !== null) {
            return true
        } else if (querySnapshot === null) {
            Crashlytics.log(
                    Log.ERROR,
                    RemoteRepository::class.java.simpleName,
                    "QueryDocumentSnapshot is null"
            )
            return true
        }
        return false
    }

    private fun shouldReturn(emitter: FlowableEmitter<*>, querySnapshot: QuerySnapshot): Boolean {
        if (emitter.isCancelled) {
            return true
        }

        if (recentLocalChanges) {
            recentLocalChanges = recentLocalChanges.not()
            return true
        } else if (fromLocalCache(querySnapshot)) {
            recentLocalChanges = true
            return false
        }
        return false
    }

    /**
     * This payload is from the local cache, post a local change.
     */
    private fun fromLocalCache(querySnapshot: QuerySnapshot) =
            querySnapshot.metadata.hasPendingWrites()
}
